src/_common/user/follow/AppUserFollowButton.vue

```diff
@@ -0,0 +1,132 @@
+<script lang="ts" setup>
+import { computed, PropType, toRefs } from 'vue';
+import { trackUserFollow, UserFollowLocation } from '../../analytics/analytics.service';
+import { vAppAuthRequired } from '../../auth/auth-required-directive';
+import AppButton from '../../button/AppButton.vue';
+import { formatNumber } from '../../filters/number';
+import { showErrorGrowl } from '../../growls/growls.service';
+import { ModalConfirm } from '../../modal/confirm/confirm-service';
+import { useCommonStore } from '../../store/common-store';
+import { vAppTooltip } from '../../tooltip/tooltip-directive';
+import AppTranslate from '../../translate/AppTranslate.vue';
+import { $gettext } from '../../translate/translate.service';
+import { followUser, unfollowUser, User } from '../user.model';
+
+const props = defineProps({
+	user: {
+		type: Object as PropType<User>,
+		required: true,
+	},
+	location: {
+		type: String as PropType<UserFollowLocation>,
+		required: true,
+	},
+	overlay: {
+		type: Boolean,
+	},
+	block: {
+		type: Boolean,
+	},
+	sm: {
+		type: Boolean,
+	},
+	hideCount: {
+		type: Boolean,
+	},
+	disabled: {
+		type: Boolean,
+	},
+});
+
+const emit = defineEmits({
+	follow: () => true,
+	unfollow: () => true,
+});
+
+const { user, hideCount, location } = toRefs(props);
+const { user: sessionUser } = useCommonStore();
+
+const shouldShow = computed(() => {
+	if (!sessionUser.value) {
+		return true;
+	}
+	if (sessionUser.value.id !== user.value.id && !user.value.hasAnyBlock) {
+		return true;
+	}
+	return false;
+});
+
+const badge = computed(() => {
+	return !hideCount.value && user.value.follower_count
+		? formatNumber(user.value.follower_count)
+		: '';
+});
+
+const tooltip = computed(() => {
+	return !user.value.is_following
+		? $gettext(`Follow this user to get their posts in your feed!`)
+		: undefined;
+});
+
+async function onClick() {
+	if (!sessionUser.value) {
+		return;
+	}
+
+	let failed = false;
+	if (!user.value.is_following) {
+		try {
+			await followUser(user.value);
+			emit('follow');
+		} catch (e) {
+			failed = true;
+			showErrorGrowl($gettext(`Something has prevented you from following this user.`));
+		} finally {
+			trackUserFollow(true, { failed, location: location.value });
+		}
+	} else {
+		try {
+			const result = await ModalConfirm.show(
+				$gettext(`Are you sure you want to unfollow this user?`),
+				$gettext(`Unfollow user?`)
+			);
+
+			if (!result) {
+				return;
+			}
```

Comment:
Check for the result outside of the try{}catch{}finally{} block, otherwise we'll track an unfollow when the user doesn't confirm.

Reply:
wow, i didn't actually know that finally ALWAYS runs, even if you early return out from it